packages/salesforcedx-vscode-core/src/commands/baseDeployCommand.ts

```diff
@@ -55,44 +54,70 @@ export abstract class BaseDeployExecutor extends SfdxCommandletExecutor<
     });
 
     execution.processExitSubject.subscribe(async exitCode => {
-      const telemetry = new TelemetryBuilder();
-      let success = false;
-      try {
-        BaseDeployExecutor.errorCollection.clear();
-        if (stdOut) {
-          const deployParser = new ForceDeployResultParser(stdOut);
-          const errors = deployParser.getErrors();
-          if (errors && !deployParser.hasConflicts()) {
-            channelService.showChannelOutput();
+      await this.exitProcessHandlerDeploy(
+        exitCode,
+        stdOut,
+        workspacePath,
+        execFilePathOrPaths,
+        execution,
+        startTime,
+        cancellationToken,
+        cancellationTokenSource
+      );
+    });
+  }
+
+  protected async exitProcessHandlerDeploy(
+    exitCode: number | undefined,
+    stdOut: string,
+    workspacePath: string,
+    execFilePathOrPaths: string,
+    execution: CliCommandExecution,
+    startTime: [number, number],
+    cancellationToken: vscode.CancellationToken | undefined,
+    cancellationTokenSource: vscode.CancellationTokenSource
+  ): Promise<void> {
+    if (exitCode === 0 && this.getDeployType() === DeployType.Push) {
+      const pushResult = JSON.parse(stdOut);
+      this.updateCache(pushResult);
+    }
+
+    const telemetry = new TelemetryBuilder();
+    let success = false;
+    try {
+      BaseDeployExecutor.errorCollection.clear();
+      if (stdOut) {
+        const deployParser = new ForceDeployResultParser(stdOut);
+        const errors = deployParser.getErrors();
+        if (errors && !deployParser.hasConflicts()) {
+          channelService.showChannelOutput();
+          if (workspacePath && execFilePathOrPaths) {
             handleDiagnosticErrors(
               errors,
               workspacePath,
               execFilePathOrPaths,
               BaseDeployExecutor.errorCollection
             );
-          } else {
-            success = true;
           }
-          this.outputResult(deployParser);
-        }
-      } catch (e) {
-        BaseDeployExecutor.errorCollection.clear();
-        if (e.name !== 'DeployParserFail') {
-          e.message =
-            'Error while creating diagnostics for vscode problem view.';
+        } else {
+          success = true;
         }
-        telemetryService.sendException(e.name, e.message);
-        console.error(e.message);
+        this.outputResult(deployParser);
       }
-      telemetry.addProperty('success', String(success));
-      this.logMetric(
-        execution.command.logName,
-        startTime,
-        telemetry.build().properties
-      );
-      await DeployQueue.get().unlock();
-    });
-
+    } catch (e) {
+      BaseDeployExecutor.errorCollection.clear();
+      if (e.name !== 'DeployParserFail') {
+        e.message = 'Error while creating diagnostics for vscode problem view.';
+      }
+      telemetryService.sendException(e.name, e.message);
+      console.error(e.message);
```

Comment:
I know you are just moving this code, but it weird we don't tell the user anything if it goes sideways outside of the dev console.

Reply:
should this return or throw?  Looks like with the current flow we're sending telemetry for both an exception and success.